### ISSUE TITLE
`sudo` access? or `user` based access?

### ISSUE THREAD

**Issue by icyflame** [athityakumar/ruby-notificator#1]
Hey, nice idea.

> Secure login data with sudo access only

This doesn't seem to fit, I think that you should rather have "silos", with each silo consisting a set of accounts which were authenticated for using (say) a 6-digit code, instead of giving only the super user access to everything.

**Comment by athityakumar** [comment]
Thanks! 😀 

The base idea is that the username & password (not API keys) will be stored in the `secrets/` folder as json files like github.json , gmail.json. Multiple accounts can be accounted for as JSONs are parsed and read as lists... But the json files should be having root-only access with chmod 700, so that others (read, college mates who typically change your profile picture when you leave your Facebook account open) can't misuse by having access to those json files... 

Also, unique 6-digit PINs for each account can be tiring for the user too... Rather, just have the program to be run only by `sudo ruby main.rb`

**Comment by icyflame** [comment]
@athityakumar 

Hmm, well. Storing passwords on disk, in plain text is _really_ not a good idea. Rather, you should get the password from the user and exchange it for an O-Auth token that you can get from the GitHub API, and store only those.

There will be only keys, and not usernames and passwords, stored on disk, so stealing them is of no use at all. (since these keys will give you only read-only access to the notifications, there's nothing someone can do with them even after stealing them).

Further more, it is pretty rare that I would trust a ruby script to be run as `sudo`. An attacker could simply replace `main.rb` with the line `rm -rf /` and that would be the end of things for the user.

TL; DR Don't ask the user to do `sudo` unless absolutely necessary. 

This is followed by bundler, among a LOT of other gems, and is really nice because it asks for sudo only if we ask it to install to the system gemset instead of `vendor/bundle`.

But anyway, I just wanted to get this problem into your attention. And TBH, I don't think that 6/4-digit passwords are hard to remember considering that this script is going to basically have access to a LOT of data.

**Comment by athityakumar** [comment]
`rm -rf /` will still be ineffective as the `secrets/` will still have mod access of 700. And to chmod / chown, again sudo password will be required. So, I don't think sudo can be by-passed. Right?

Yeah, maybe trading the credentials for API keys would be better. But even if PINs are used, anyway it'll still be on the disk and sudo seemed better security to me.

Of course, if someone knows how to boot into grub & reset the sudo password - there is no way anyone can stop him. 😉

**Comment by icyflame** [comment]
@athityakumar I don't know what you mean. Why would `sudo rm -rf /` be ineffective? As soon as the sudo password is entered, the system will let the command run. `sudo` has access to all files, everywhere, so, it can very well remove everything.

Hence, either the Ruby script itself should be owned by the super user and no one else should have anything but read access to it, which is the only way I think that the system can be secure, even to some extent.

**Comment by athityakumar** [comment]
I mean, sudo rm -rf / does require sudo password in the first place, right?
So none other than the actual user (who knows sudo password) will be able
to use the credentials right?

**Comment by icyflame** [comment]
@athityakumar :

> An attacker could simply replace main.rb with the line rm -rf / and that would be the end of things for the user.

Attacker changes the script, but the ACTUAL sudo user doesn't know. So, the script itself would need to be `chown`ed and `chmod`ed to let only the sudo user edit it. Not the best setup, is my point.

**Comment by athityakumar** [comment]
Hmm... Maybe storing these credentials into SQL tables with PINs would be better then?

**Comment by icyflame** [comment]
Um, maybe. But storing oAuth tokens would be the best.

I did have to do this selection for an NPM module which had github integration, and it was fairly insensitive data so, I just selected to take the password from the user, exchanged it for an OAuth token and then stored the token inside ~/.config. ~/.config is setup to have permissions of 600, read-write for owner only, nothing for everyone else. Fairly safe.

In your case, I would say store OAuth inside a 600 folder, and the 4-digit PIN also in there.

This way, the only person who can mess with you is someone with `sudo` access to the machine, and frankly, if an attacker has sudo access to the machine, you are pretty much screwed anyway. Your Facebook notifications are the last thing you need to worry about!

**Comment by athityakumar** [comment]
Got it! 😀
